### ISSUE TITLE
Proposal: Permit variable declarations under disjunctive patterns

### ISSUE THREAD

**Issue by alrz** [dotnet/csharplang#4018]
# Permit variable declarations under disjunctive patterns

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Allow pattern variables to be declared in different mutually exclusive patterns. This is the part of [pattern-matching proposal](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/patterns3.md) that did not make it into C# 9.0.

## Motivation
[motivation]: #motivation

The feature would reduce code duplication where switch expression arms or switch section bodies could be shared for a number of patterns.
```cs
switch (e)
{
    case (int i, 0):
        Use(i); break;
    case (0, int i):
        Use(i); break;
}

switch (e)
{
    case (int i, 0):
    case (0, int i):
        Use(i); break;
}
```
```cs
e switch
{
    (int i, 0) => i,
    (0, int i) => i,
}

e switch
{
    (int i, 0) or (0, int i) => i,
}
```

Today we can workaround this via breaking each pattern to separate arms or sections. 

For `is` expressions that is even worse as we need to use a different name for each branch.
```cs
if (e is (int i, 0) or (0, int i))
  Use(i);

if (e is (int i, 0))
  Use(i);
else if (e is (0, int j))
  Use(j);
```

## Detailed design
[design]: #detailed-design

For this to work, we basically require each mutually exclusive pattern to define the same set of variables with the same types, otherwise they would not be definitely assigned in that particular code path. This includes `or` patterns and each case of a switch section.

As a consequence, a pattern variable may be definitely assigned in a `when` clause but not in the switch section body, for example.

```cs
  case (int x, 0) a when Use(x, a): // ok
  case (0, int x) b when Use(x, b): // ok
     Use(x); // ok
     Use(a); // error; not definitely assigned
     Use(b); // error; not definitely assigned
     break;
```

Pattern variables may be recursively multiply defined on each side:

```cs
if (e is { A: (int x, 0) or (0, int x) } or
         { B: (int x, 0) or (0, int x) })
{
    Use(x);
}  
```

Under the hood, it is the same local being assigned in each pattern.

Since we require an identical type for each declaration, the narrowed type would not be affected by this change.

## Further considerations

We could extend the rules to all expression variables since there is no good workaround for it at the moment (https://github.com/dotnet/csharplang/issues/3740#issuecomment-665151696).

```cs
if (a is int x ||
    b is int x)
{
    Use(x); 
}
```

The repeated pattern basically defines a local with multiple declarations in it's scope. So it should be possible to apply the same scoping rules when we are analyzing expression variables anywhere in the expression boundaries, not only inside each pattern individually.

## Design meetings

None.

**Comment by alrz** [comment]
cc @333fred

**Comment by CyrusNajmabadi** [comment]
I will champ ion this.

**Comment by HaloFour** [comment]
I imagine we can crib a lot of the rules/spec from F# since they offer this functionality?

**Comment by alrz** [comment]
@HaloFour Definitely. F# specification on the subject did help a lot already.

> `Both patterns must bind the same set of variables with the same types`

**Comment by 333fred** [comment]
Lol @CyrusNajmabadi I had already told alrz if champion this. It's mine!

**Comment by CyrusNajmabadi** [comment]
WFM :)  I was championing in case no one else picked it up.  But if you want it, go for it :)

**Comment by MadsTorgersen** [comment]
I want us to address this scenario! I am a bit concerned about a feature that has multiple declarations of the same variable, so I would want to consider alternatives that didn't do that.

For instance, maybe we have a pattern that can match into an existing variable if the type is right. E.g. (just thinking out loud here):

``` c#
e switch
{
    (int i, 0) or (0, i) => i,
}
```
Where specifying an existing variable matches _into_ that variable. Perhaps we would only allow it if the variable is not already assigned, to avoid pattern matching having side effects.

I'm not saying this is better, but it has different trade-offs, and we should consider our options.

**Comment by alrz** [comment]
Some thoughts on that syntax:

1. It's not clear that we're assigning to a variable or matching a constant. The former is something that we might want to consider at some point but I don't think a sole identifier is suitable. I think it should be reserved for when we want to possibly permit matching with non-constatnt readonly variables.
2. We depend on the order in which the variable is declared first, so reordering patterns around won't be free.
3. This doesn't seemlessly apply if we want to do this outside pattern boundaries (see the "further considerations" section)

**Comment by HaloFour** [comment]
> Where specifying an existing variable matches into that variable.

Would that mean enabling using existing variables in pattern variables in general?

```cs
int i;
e switch {
    (i, 0) or (0, i) => i
}
```

**Comment by alrz** [comment]
What if I want to *match* instead of assign?
```cs
readonly int i = 0;
e switch {
    (i, 0) or (0, i) => i
}
```
I've suggested `out i` for assigning existing variables before but I think it would be mainly useful with user-defined positional patterns.

**Comment by HaloFour** [comment]
Exactly, that syntax wades into the territory of requests to allow existing variables to be used in patterns instead of just constants, as well as being able to assign patterns into existing variables.

**Comment by alrz** [comment]
>as well as being able to assign patterns into existing variables

And to be clear, the said variables are usually defined outside of the pattern itself and likely unassigned (e.g. out parameters)

I'd even say that if it ever become a thing, it should be disallow to reassign pattern variables, that'll make the pattern itself difficult to reason about and to hard to follow where such variables are assigned and possibly reassigned.

So I favor "multiply declared variables" syntax where we explicitly re-declare variables, rather than implicitly re-assigning them.

**Comment by alrz** [comment]
From LDM:

>We need a rule that says when you are allowed to redeclare existing variables. It needs to cover multiple switch case labels, while also not permitting things declared outside the switch label to be redeclared.


It seems like to allow this we only need to relax the existing scoping rules around pattern variables to enable multi-declarations and handle definite-assignment accordingly. i.e. `ExpressionVariableBinder` would need to handle pattern variables differently.

In other words, the rules that flag those name conflicts are already there, we only need to relax those existing scoping rules.


>How identical do the types need to be? Are nullability differences permitted? ie, are (object?, object) and (object, object?) the same for the purposes of this feature? It seems like they may have to be.

Pattern variables start with not-null state, so I think this is a moot point.

**Comment by 333fred** [comment]
> In other words, the rules that flag those name conflicts are already there, we only need to relax those existing scoping rules. 

We still need the specese. 

> Pattern variables start with not-null state, so I think this is a moot point.

Consider:
```cs
if (o is (not null, null) a or (null, not null) a) {} 
````

Also remember the `var` pattern.

**Comment by Thaina** [comment]
I want to add that I wish I could reassign variable. So that I could chain it in if block

Instead of

```C#
var a = GetFrom0();
if(a == null)
    a = GetFrom1();
if(a == null)
    a = GetFrom2();
if(a == null)
    return;

// DoSomething with a
```

I could just

```C#
if(GetFrom0() is var a || GetFrom1() is {} a || GetFrom2() is {} a)
{
    // DoSomething with a
}
```

**Comment by spydacarnage** [comment]
> ```cs
> if(GetFrom0() is var a || GetFrom1() is {} a || GetFrom2() is {} a) // Should it require reassign to be {} only ?
> {
>     // DoSomething with a
> }
> ```

Can't you use:
```cs
if ((GetFrom0() ?? GetFrom1() ?? GetFrom2()) is var a)
{
    // DoSomething with a
}
```

**Comment by alrz** [comment]
>So that I could chain it in if block

That's the second part of the proposal where you could possibly redeclare pattern variables and they assign to the same local

> ```cs
> if (M() is int x ||
>     N() is int x)
> {
>     Use(x); 
> }
> ```

**Comment by Thaina** [comment]
@alrz Oh then sorry I miss that part

**Comment by Thaina** [comment]
@spydacarnage Not only that sometime I want to do something like this

```C#
void DoSomething(SomeObject obj)
{
    if(obj != null || GetFrom0() is {} obj || GetFrom1() is {} obj || GetFrom2() is {} obj) // reassign variable
    {
        // DoSomething with obj
    }
}
```
